### ISSUE TITLE
Enzyme TA expectations/issues

### ISSUE THREAD

**Issue by ZuseZ4** [EnzymeAD/Enzyme#2006]
Splitting this out of the Rust discussion, since it runs into Enzyme core issues.

1) Enzyme will often fail to run TA (and thus differentiate) llvm-ir, if no additional type information is provided.
2) For the Rust side, I therefore started to provide TT for the outermost function getting differentiated, as well as some memset/memcpy/mem* calls. We want to potentially provide TT in even more places.
3) This now causes issues, because Rust has sufficient locations where ptr2int or int2ptr casts happen. Context is fundamentally not available when lowering, so we do not now if an int is going to get casted to a pointer (or vice versa) when creating TT. Refactoring rustc to carry enough context to figure if such a usage will happen would be too happen (and also against other ongoing efforts like querification, parallelization, and caching/inc compilation.

I don't have the capacity to work on Enzyme core, but if you can provide a way to tell Enzyme through llvm-ir metadata that ptrIntSame=1 for a specific argument, I can look at how reinterpret casts are getting lowered to llvm-ir (hoping it's not a no-op and actually get's lowered). [std::mem::transmute](https://doc.rust-lang.org/std/mem/fn.transmute.html#transmutation-between-pointers-and-integers) and `as` casts are the two things I'd look into.
In that case, we could retroactively tell Enzyme, that arguments to this cast are to be treated with ptrIntSame=1.
Not sure about Enzyme core, but I could imagine that it requires running past the first TA error to check if we have such a notification from rustc to enzyme later?
If you don't have the capacity to provide that, I would probably just set PtrIntSame=1 for every pointer and Integer in all the TT that we create. I guess this would reduce performance (?) and potentially obscure bugs in cases where we create the wrong TT or Enzyme's TA has a corresponding bug, but at least it would work. Right now I have multiple people blocked on this issue and it is a bug I see quite often, so either of the two would be ok for me, what do you think?

**Comment by wsmoses** [comment]
So you can potentially add a flag that tells enzyme to always set int and ptr as the same.

It may introduce other issues as well since now all ints may be active so I’ll let you think through the rest of pros/cons for a bit and maybe present an rfc ?

**Comment by ZuseZ4** [comment]
Reasigning, can you please say something about the local option I described above? In my opinion here Enzyme Core lacks the feature to retroactively and through llvm-ir metadata mark a specific variable as ptrintsame.
> I don't have the capacity to work on Enzyme core, but if you can provide a way to tell Enzyme through llvm-ir metadata that ptrIntSame=1 for a specific argument, I can look at how reinterpret casts are getting lowered to llvm-ir 

Yes, I described some of the issues of a brute force, global ptrintsame above (i.e. worse runtime perf and potentially catching less Enzyme core bugs), so I'd prefer to avoid that while there is a chance for Enzyme core to provide a local update info, that I could use from the Rust side.

**Comment by ZuseZ4** [comment]
Unfortunately simple as and transmute cases pass, so I can't show a mwe here: https://github.com/EnzymeAD/rustbook/commit/33a9ae3b5dbcf9b3fc86dabc371d120482b11914
But for the options that lead to store an int to a ptr, Enzyme should have a flag that allows it to retroactively change the ptrintsame setting of that variable.

**Comment by wsmoses** [comment]
At the moment int and pointer equivalence is done at an analysis wide (or use case wide) level of granularity not per variable. This is because different use cases will potentially be invalid if this is done anywhere so thinking how to do so properly at a variable level requires some thoughtful design.

Happy to look over a design if you have one, and maybe you’d be interested in presenting at the next odm?

**Comment by wsmoses** [comment]
this doesnt seem like an actionable issue, closing
